### PR TITLE
Add errorMessage in taskStatus for task failures in middleManagers/indexers/peons

### PR DESCRIPTION
**Pull request by jihoonson**
### Description

This PR is a follow-up of https://github.com/apache/druid/pull/11419 and fills the error message in `TaskStatus` properly when the task fails in the middleManager, indexer, or peon. It includes a minor refactoring for `ForkingTaskRunner` to make it more testable. Other than that, all changes are quite straight forward.

While I was making this PR, I remembered an issue with `TaskRunnerListener` that sometimes a wrong taskStatus is fed even though `TaskRunnerListener` is almost not in use. I created https://github.com/apache/druid/issues/11445 to track this issue.

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/ForkingTaskRunner.java [hunk 1]
```diff
@@ -364,38 +362,19 @@ public TaskStatus call()
...
+                        LOGGER.info("Process exited successfully for task: %s", task.getId());
                         // Process exited successfully
                         status = jsonMapper.readValue(statusFile, TaskStatus.class);
                       } else {
+                        LOGGER.error("Process exited with code[%d] for task: %s", exitCode, task.getId());
                         // Process exited unsuccessfully
-                        status = TaskStatus.failure(task.getId());
+                        status = TaskStatus.failure(
+                            task.getId(),
+                            "Task execution process exited unsuccessfully. See middleManager logs for more details."
```
Comment (abhishekagarwal87) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/ForkingTaskRunner.java:376]: do you want to add the exit code in the message here?

**Review by abhishekagarwal87 — APPROVED** [review]
Minor comments. LGTM otherwise.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/ForkingTaskRunner.java (hunk 1 above)
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/ForkingTaskRunner.java]: It seems reasonable to include it. Unlike other exceptions thrown in Druid, middleManager logs will probably not provide much information about the process failure. :+1:

File: indexing-service/src/test/java/org/apache/druid/indexing/overlord/ForkingTaskRunnerTest.java [hunk 2]
```diff
@@ -145,12 +162,192 @@ public void testMaskedIterator()
...
+      {
+        // Emulate task process failure
+        return 1;
+      }
+    };
+
+    final TaskStatus status = forkingTaskRunner.run(NoopTask.create()).get();
+    Assert.assertEquals(TaskState.FAILED, status.getStatusCode());
+    Assert.assertEquals(
+        "Task execution process exited unsuccessfully. See middleManager logs for more details.",
```
Comment (loquisgon) [inline indexing-service/src/test/java/org/apache/druid/indexing/overlord/ForkingTaskRunnerTest.java:235]: Now that you added the exit code to the message you need to update this message as well

Comment (loquisgon) [reply indexing-service/src/test/java/org/apache/druid/indexing/overlord/ForkingTaskRunnerTest.java:235]: "Task execution process exited unsuccessfully with code[1]. See middleManager logs for more details."

**Review by loquisgon — COMMENTED** [review]
LGTM other than the test coverage failure & adding the exit code to the error msg

File: indexing-service/src/test/java/org/apache/druid/indexing/overlord/ForkingTaskRunnerTest.java (hunk 2 above)
Comment (jihoonson) [reply indexing-service/src/test/java/org/apache/druid/indexing/overlord/ForkingTaskRunnerTest.java]: Oops thanks :+1:

Comment (loquisgon) [reply indexing-service/src/test/java/org/apache/druid/indexing/overlord/ForkingTaskRunnerTest.java]: LGTM

**Comment by jihoonson** [comment]
@abhishekagarwal87 @loquisgon thanks for the review :+1:

**Pull request merged**